// @sudoku/sudoku.js
import solve from '@mattflow/sudoku-solver';
import { BOX_SIZE, GRID_LENGTH, SUDOKU_SIZE, GRID_COORDS } from './constants';
import { getSudoku } from 'fake-sudoku-puzzle-generator';

export class Sudoku {
  constructor(grid = null) {
    // 防御性拷贝：确保外部修改原数组不会影响内部
    this.grid = grid 
      ? grid.map(row => [...row]) 
      : Array.from({ length: 9 }, () => Array(9).fill(0));
  }

  getGrid() {
    return this.grid;
  }

  /**
   * 核心修复：兼容 (row, col, value) 和 ({row, col, value}) 两种写法
   */
  guess(rowOrObject, col, value) {
    let r, c, v;
    if (typeof rowOrObject === 'object' && rowOrObject !== null) {
      r = rowOrObject.row;
      c = rowOrObject.col;
      v = rowOrObject.value;
    } else {
      r = rowOrObject;
      c = col;
      v = value;
    }
    
    // 安全检查，防止 grid 未初始化或索引越界
    if (this.grid && this.grid[r]) {
      this.grid[r][c] = v;
    }
  }

  toJSON() {
    return { grid: this.grid.map(row => [...row]) };
  }

  toString() {
    return this.grid.map(row => row.join(',')).join('\n');
  }

  static generate(difficulty = 'easy') {
    const raw = getSudoku(difficulty);
    const processed = raw.map(row => row.map(v => v === null ? 0 : v));
    return new Sudoku(processed);
  }

  solve() {
    const solutionArray = solve(this.grid.flat().join(''), {
      outputArray: true,
      hintCheck: false
    });
    let solutionGrid = Array.from({ length: 9 }, () => Array(9).fill(0));
    for (let cell = 0; cell < 81; cell++) {
      const [row, col] = GRID_COORDS[cell];
      solutionGrid[row][col] = solutionArray[cell];
    }
    return new Sudoku(solutionGrid);
  }

  clone() {
    return new Sudoku(this.grid.map(row => [...row]));
  }
}
