// @sudoku/sudoku.js
import solve from '@mattflow/sudoku-solver';
// 注意：这里去掉了别名，改用相对路径
import { BOX_SIZE, GRID_LENGTH, SUDOKU_SIZE, GRID_COORDS } from './constants';
import { getSudoku } from 'fake-sudoku-puzzle-generator';

export class Sudoku {
  constructor(grid = null) {
    this.grid = grid || Array.from({ length: 9 }, () => Array(9).fill(0));
  }

  // 契约测试必填：获取棋盘数据
  getGrid() {
    return this.grid;
  }

  // 契约测试必填：填数接口
  guess(row, col, value) {
    this.grid[row][col] = value;
  }

  static generate(difficulty = 'easy') {
    const raw = getSudoku(difficulty);
    const processed = raw.map(row => row.map(v => v === null ? 0 : v));
    return new Sudoku(processed);
  }

  solve() {
    const solutionArray = solve(this.grid.flat().join(''), {
      outputArray: true,
      hintCheck: false
    });

    let solutionGrid = Array.from({ length: 9 }, () => Array(9).fill(0));
    for (let cell = 0; cell < 81; cell++) {
      const [row, col] = GRID_COORDS[cell];
      solutionGrid[row][col] = solutionArray[cell];
    }
    return new Sudoku(solutionGrid);
  }

  clone() {
    return new Sudoku(this.grid.map(row => [...row]));
  }
}
